File: src/rag-module/ingestion.ts
import { buildExactHash, buildTemplateHash } from "./fingerprint";
import type { IngestionFile } from "./types";

type ParsedQuestion = {
  stem: string;
  options: string[];
  answer?: string;
  explanation?: string;
  sourceQuestionNo?: number;
};

function cleanupText(input: string): string {
  return input.replace(/\r/g, "").replace(/\t/g, " ").replace(/\u00a0/g, " ").trim();
}

function parseOptions(block: string): string[] {
  const lines = cleanupText(block).split("\n");
  const options: string[] = [];
  const optionRe = /^\s*([A-D])[).:\-]\s*(.+)$/i;
  for (const line of lines) {
    const m = line.match(optionRe);
    if (m) options.push(m[2].trim());
  }
  return options;
}

export function parseQuestionsFromPlainText(content: string): ParsedQuestion[] {
  const text = cleanupText(content);
  if (!text) return [];

  const boundaryRe = /(?:^|\n)\s*(?:question\s*\d+[\).:]|\d+[\).:])\s+/gi;
  const chunks = text.split(boundaryRe).map((x) => x.trim()).filter(Boolean);
  const candidates = chunks.length > 1 ? chunks : [text];
  const out: ParsedQuestion[] = [];

  for (let i = 0; i < candidates.length; i++) {
    const chunk = candidates[i];
    const answer = chunk.match(/\banswer\s*[:\-]\s*([A-D])\b/i)?.[1]?.toUpperCase();
    const explanation = chunk.match(/\bexplanation\s*[:\-]\s*([\s\S]*)$/i)?.[1]?.trim();
    const stem = chunk.split(/\n\s*[A-D][).:\-]\s+/i)[0]?.trim();
    const options = parseOptions(chunk);

    if (!stem) continue;
    out.push({
      stem,
      options,
      answer,
      explanation,
      sourceQuestionNo: i + 1,
    });
  }

  return out;
}

export function parseQuestionsFromFile(file: IngestionFile): ParsedQuestion[] {
  // Stageable parser: current version expects extracted text, but keeps per-file entry point.
  return parseQuestionsFromPlainText(file.content);
}

export function buildFingerprints(stem: string, options: string[], answer?: string): { exactHash: string; templateHash: string } {
  return {
    exactHash: buildExactHash(stem, options, answer),
    templateHash: buildTemplateHash(stem),
  };
}

File: rag_python/ingestion.py
from __future__ import annotations

import re


def parse_questions_from_plain_text(content: str) -> list[dict]:
    text = content.replace("\r", "").strip()
    if not text:
        return []
    boundary = re.compile(r"(?:^|\n)\s*(?:question\s*\d+[\).:]|\d+[\).:])\s+", re.IGNORECASE)
    chunks = [c.strip() for c in boundary.split(text) if c.strip()]
    if len(chunks) == 1:
        chunks = [text]

    out: list[dict] = []
    for idx, chunk in enumerate(chunks, 1):
        answer_m = re.search(r"\banswer\s*[:\-]\s*([A-D])\b", chunk, re.IGNORECASE)
        expl_m = re.search(r"\bexplanation\s*[:\-]\s*([\s\S]*)$", chunk, re.IGNORECASE)
        stem = re.split(r"\n\s*[A-D][).:\-]\s+", chunk, maxsplit=1, flags=re.IGNORECASE)[0].strip()
        options = []
        for line in chunk.split("\n"):
            m = re.match(r"^\s*[A-D][).:\-]\s*(.+)$", line.strip(), re.IGNORECASE)
            if m:
                options.append(m.group(1).strip())
        if stem:
            out.append(
                {
                    "stem": stem,
                    "options": options,
                    "answer": answer_m.group(1).upper() if answer_m else None,
                    "explanation": expl_m.group(1).strip() if expl_m else None,
                    "source_question_no": idx,
                }
            )
    return out


def parse_questions_from_file(file_row: dict) -> list[dict]:
    return parse_questions_from_plain_text(str(file_row.get("content", "")))
